data_steward/app_identity.py

```diff
@@ -12,7 +12,11 @@ def get_application_id():
     # NOTE: Google interchangeably refers to this identifier as application_id or project_id
     project_id = os.environ.get(PROJECT_ID)
 
-    if project_id:
-        return project_id
-    else:
-        raise RuntimeError('{} is not set.  Set and retry.'.format(PROJECT_ID))
+    # ensure project id is a non-empty string
+    if isinstance(project_id, str):
+        project_id = project_id.strip()
+        if project_id is not '':
+            return project_id
```

Comment:
nit:  This works, but I'm not convinced it's necessary.  Environment variables are set when that first space is met.  So below, `FOO` is set to an empty string, `BAR` is good, and `BAZ= bad` results in a command not found error.  So, I don't think it's possible to have `FOO=<multiple blanks>` from an environment variable.  Also, environment variables are always going to give us [strings](https://docs.python.org/3/library/os.html#os.environ).
```
(env37) [Thu Apr 08 11:50:13] [~/Projects/curation/data_steward] lbiggers $ FOO=     BAR=good BAZ= bad python
-bash: bad: command not found
(env37) [Thu Apr 08 11:50:31] [~/Projects/curation/data_steward] lbiggers $ FOO=     BAR=good BAZ=bad  python
Python 3.7.2 (default, Feb 12 2019, 08:15:36) 
[Clang 10.0.0 (clang-1000.11.45.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> os.environ.get('FOO')
''
>>> os.environ.get('BAR')
'good'
>>> os.environ.get('BAZ')
'bad'
>>> 
```
The only way this could get messed up is if someone starts injecting environment variables in through python code.  Which could happen, but it **_really_** shouldn't.  If somebody is injecting keys/values, I would argue they should be injecting them into a [Namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) not the environment dictionary.
```
>>> import os
>>> os.environ['GOOF']='     '
>>> foo(os.environ.get('GOOF'))
bar is :     
```
Finally, an empty string receives the same Truthy/Falsy treatment as None and zero.
```
(env37) [Thu Apr 08 11:54:56] [~/Projects/curation/data_steward] lbiggers $ python
Python 3.7.2 (default, Feb 12 2019, 08:15:36) 
[Clang 10.0.0 (clang-1000.11.45.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> def foo(bar):
...   if bar:
...     print(f'bar is :{bar}')
...   else:
...     print('bar is falsy')
... 
>>> foo('')
bar is falsy
>>> foo(None)
bar is falsy
>>> foo(0)
bar is falsy
>>> foo('  ')
bar is :  
>>> # but the above can't happen with true environment variables
... 
>>> foo('23')
bar is :23
>>> foo(23)
bar is :23
```

Reply:
> nit: This works, but I'm not convinced it's necessary. Environment variables are set when that first space is met. So below, `FOO` is set to an empty string, `BAR` is good, and `BAZ= bad` results in a command not found error. So, I don't think it's possible to have `FOO=<multiple blanks>` from an environment variable. Also, environment variables are always going to give us [strings](https://docs.python.org/3/library/os.html#os.environ).
> 
> ```
> (env37) [Thu Apr 08 11:50:13] [~/Projects/curation/data_steward] lbiggers $ FOO=     BAR=good BAZ= bad python
> -bash: bad: command not found
> (env37) [Thu Apr 08 11:50:31] [~/Projects/curation/data_steward] lbiggers $ FOO=     BAR=good BAZ=bad  python
> Python 3.7.2 (default, Feb 12 2019, 08:15:36) 
> [Clang 10.0.0 (clang-1000.11.45.5)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import os
> >>> os.environ.get('FOO')
> ''
> >>> os.environ.get('BAR')
> 'good'
> >>> os.environ.get('BAZ')
> 'bad'
> >>> 
> ```
>
> The only way this could get messed up is if someone starts injecting environment variables in through python code. Which could happen, but it **_really_** shouldn't. If somebody is injecting keys/values, I would argue they should be injecting them into a [Namespace](https://docs.python.org/3/tutorial/classes.html#python-scopes-and-namespaces) not the environment dictionary.
> 
> ```
> >>> import os
> >>> os.environ['GOOF']='     '
> >>> foo(os.environ.get('GOOF'))
> bar is :     
> ```
> 
> Finally, an empty string receives the same Truthy/Falsy treatment as None and zero.
> 
> ```
> (env37) [Thu Apr 08 11:54:56] [~/Projects/curation/data_steward] lbiggers $ python
> Python 3.7.2 (default, Feb 12 2019, 08:15:36) 
> [Clang 10.0.0 (clang-1000.11.45.5)] on darwin
> Type "help", "copyright", "credits" or "license" for more information.
> >>> def foo(bar):
> ...   if bar:
> ...     print(f'bar is :{bar}')
> ...   else:
> ...     print('bar is falsy')
> ... 
> >>> foo('')
> bar is falsy
> >>> foo(None)
> bar is falsy
> >>> foo(0)
> bar is falsy
> >>> foo('  ')
> bar is :  
> >>> # but the above can't happen with true environment variables
> ... 
> >>> foo('23')
> bar is :23
> >>> foo(23)
> bar is :23
> ```

I hear you on testing the "falsey" nature of the retrieved variable, however few small corrections:

1. it is very possible to have an empty string or space-prefix / suffixed envvar: 
    ```shell
    carbone@MANUS:~$ export PADDING='   padded    '
    dcarbone@MANUS:~$ export NO_PADDING='nopadding'
    dcarbone@MANUS:~$ printenv | grep PADDING
    PADDING=   padded    
    NO_PADDING=nopadding
    ```
2. when undefined, the return value from `os.getenv()` is actually `None`, not an empty str, and I did not want to try calling `.strip()` off a non-str object:
    ```python
    import os

    if __name__ == '__main__':
        print(os.getenv('nobody knows the trouble i seeeeeee').strip())
    ```
    the above errors with `AttributeError: 'NoneType' object has no attribute 'strip'`
    * The same goes when utilizing the `.get()` method directly off any dict type, which is ultimately what `os.environ` is
    * We could remediate this by using `os.getenv('{envname}', '')`, at which point we would get an empty string
        ```shell
        import os

        if __name__ == '__main__':
            print(os.getenv('nobody knows the trouble i seeeeeee', '').strip())
        ```

I think an acceptable update would be:

```python
    project_id = os.getenv.get(PROJECT_ID, '').strip()

    # ensure project id is a non-empty string
    if project_id:
        return project_id

    # in all other cases, OBJECTION!
    raise RuntimeError(f'{PROJECT_ID} is not set.  Set and retry.')
```

What do you think?

Reply:
You're right.   I forgot about the possibility of having spaces with quoted environment values.
I did know that `.get()` returned `None` for an undefined dictionary key.  The previous code counted on its falsy nature.  I prefer using the dictionary default value when you have a desired default type.  I like the code above.

Reply:
Ah, sure, I should have enhanced my reply by saying:

> With the need to strip the envvar in mind, i need to ensure its a string before calling .strip()...

etc.  Did not mean to come across as preachy :)